File: Design-Patterns-Using-Ruby/Mediator/MediatorPattern1.rb
# Mediator interface
class Mediator
    def add_colleague(colleague)
        raise NotImplementedError, 'Subclasses must override this method'
    end
  
    def send_message(message, colleague_id)
        raise NotImplementedError, 'Subclasses must override this method'
    end
end

# ConcreteMediator class
class ConcreteMediator < Mediator
    def initialize
        @colleagues = {}
    end

    def add_colleague(colleague)
        @colleagues[colleague.id] = colleague
    end

    def send_message(message, colleague_id)
        puts "Mediator pass Message : #{message}"
        @colleagues[colleague_id].receive(message)
    end
end

# Colleague abstract class
class Colleague
    attr_reader :id

    def initialize(mediator, id)
        @mediator = mediator
        @id = id
    end

    def send(message, to)
        puts "#{id} Sent Message : #{message}"
        @mediator.send_message(message, to)
    end

    def receive(message)
        puts "#{id} Received Message #{message}"
    end
end

# ConcreteColleague1 class
class ConcreteColleague1 < Colleague
    def initialize(mediator)
        super(mediator, 'First')
    end
end

# ConcreteColleague2 class
class ConcreteColleague2 < Colleague
    def initialize(mediator)
        super(mediator, 'Second')
    end
end
  
# Client code
mediator = ConcreteMediator.new
first = ConcreteColleague1.new(mediator)
mediator.add_colleague(first)
second = ConcreteColleague2.new(mediator)
mediator.add_colleague(second)
first.send('Hello, World!', 'Second')
=begin 
First Sent Message : Hello, World!
Mediator pass Message : Hello, World!
Second Received Message Hello, World!
=end

File: Design-Patterns-Using-Kotlin/Mediator/MediatorPattern1.kt
interface Mediator {
    fun addColleague(colleague: Colleague)
    fun sendMessage(message: String, colleagueId: String)
}

class ConcreteMediator : Mediator {
    private val colleagues: MutableMap<String, Colleague> = HashMap()

    override fun addColleague(colleague: Colleague) {
        colleagues[colleague.id] = colleague
    }

    override fun sendMessage(message: String, colleagueId: String) {
        println("Mediator pass Message : $message")
        colleagues[colleagueId]!!.receive(message)
    }
}

abstract class Colleague(protected var mediator: Mediator, var id: String) {
    abstract fun send(message: String, to: String)

    abstract fun receive(message: String)
}

class ConcreteColleague1(mediator: Mediator) : Colleague(mediator, "First") {
    override fun send(message: String, to: String) {
        println("$id Sent Message : $message")
        mediator.sendMessage(message, to)
    }

    override fun receive(message: String) {
        println("$id Received Message $message")
    }
}

class ConcreteColleague2(mediator: Mediator) : Colleague(mediator, "Second") {
    override fun send(message: String, to: String) {
        println("$id Sent Message : $message")
        mediator.sendMessage(message, to)
    }

    override fun receive(message: String) {
        println("$id Received Message $message")
    }
}

// Client Code
fun main() {
    val mediator = ConcreteMediator()
    val first = ConcreteColleague1(mediator)
    mediator.addColleague(first)
    val second = ConcreteColleague2(mediator)
    mediator.addColleague(second)

    first.send("Hello, World!", "Second")
}

/*
First Sent Message : Hello, World!
Mediator pass Message : Hello, World!
Second Received Message Hello, World!
*/